sidecar/dns/server.go

```diff
@@ -0,0 +1,288 @@
+	err := s.dnsServer.ListenAndServe()
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error starting DNS server")
+	}
+
+	return nil
+}
+
+// Shutdown stpos the DNS server
+func (s *Server) Shutdown() error {
+	logrus.Info("Shutting down DNS server")
+	err := s.dnsServer.Shutdown()
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error shutting down DNS server")
+	} else {
+		logrus.Info("DNS server has shutdown")
+	}
+
+	return err
+}
+
+func (s *Server) handleRequest(w dns.ResponseWriter, request *dns.Msg) {
+	response := new(dns.Msg)
+	response.SetReply(request)
+	response.Extra = request.Extra
+	response.Authoritative = true
+	response.RecursionAvailable = false
+
+	for i, question := range request.Question {
+		err := s.handleQuestion(question, request, response)
+		if err != nil {
+			logrus.WithError(err).Errorf("Error handling DNS question %d: %s", i, question.String())
+			// TODO: what should the dns response return ?
+			break
+		}
+	}
+	err := w.WriteMsg(response)
+	if err != nil {
+		logrus.WithError(err).Errorf("Error writing DNS response")
+	}
+}
+
+func (s *Server) handleQuestion(question dns.Question, request, response *dns.Msg) error {
+
+	switch question.Qclass {
+	case dns.ClassINET:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return fmt.Errorf("unsupported DNS question class: %v", dns.Class(question.Qclass).String())
+	}
+
+	switch question.Qtype {
+	case dns.TypeA:
+	case dns.TypeAAAA:
+	case dns.TypeANY:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return fmt.Errorf("unsupported DNS question type: %v", dns.Type(question.Qtype).String())
+	}
+
+	serviceInstances, err := s.retrieveServices(question, request, response)
+
+	if err != nil {
+		return err
+	}
+
+	err = s.findMatchingServices(question, request, response, serviceInstances)
+	return err
+
+}
+
+func (s *Server) retrieveServices(question dns.Question, request, response *dns.Msg) ([]*client.ServiceInstance, error) {
+	// parse query :
+	// Query format:
+	// [tag]*.<service>.<domain>.
+	numberOfLabels, isValidDomain := dns.IsDomainName(question.Name)
+	if !isValidDomain {
+		response.SetRcode(request, dns.RcodeBadName)
+		return nil, fmt.Errorf("Invalid Domain name %s", question.Name)
+	}
+
+	fullDomainRequestArray := dns.SplitDomainName(question.Name)
+	if len(fullDomainRequestArray) == 1 {
+		response.SetRcode(request, dns.RcodeNameError)
+		return nil, fmt.Errorf("service name wasn't included in domain %s", question.Name)
+
+	}
+	serviceName := fullDomainRequestArray[numberOfLabels-2]
+	tags := fullDomainRequestArray[:len(fullDomainRequestArray)-2]
+	var ServiceInstances []*client.ServiceInstance
+	var err error
+	if len(tags) == 0 {
+		ServiceInstances, err = s.config.DiscoveryClient.ListServiceInstances(serviceName)
+	} else {
+		filters := client.InstanceFilter{ServiceName: serviceName, Tags: tags}
+		ServiceInstances, err = s.config.DiscoveryClient.ListInstances(filters)
```

Comment:
This code, as is wired today, is calling the remote registry.
@tomerGolany has started working on another change-set so that this will be served from the local cache. Specifically, he is changing the `RegistryMonitor` so that it'll also implement the `github.com/amalgam8/amalgam8/registry/client.Discovery` interface, and to inject that into the dns server, instead of the existing HTTP-based discovery client.